primitives/core/src/ecdsa.rs

```diff
@@ -348,31 +357,34 @@ impl Signature {
 	/// Recover the public key from this signature and a message.
 	#[cfg(feature = "full_crypto")]
 	pub fn recover<M: AsRef<[u8]>>(&self, message: M) -> Option<Public> {
-		let message = secp256k1::Message::parse(&blake2_256(message.as_ref()));
-		let sig: (_, _) = self.try_into().ok()?;
-		secp256k1::recover(&message, &sig.0, &sig.1)
+		let message = secp256k1::Message::from_slice(&blake2_256(message.as_ref())).ok()?;
+
+		let id = secp256k1::recovery::RecoveryId::from_i32(self.0[64] as i32).ok()?;
+		let sig = secp256k1::recovery::RecoverableSignature::from_compact(&self.0[..64], id).ok()?;
+
+		secp256k1::Secp256k1::verification_only().recover(&message, &sig)
 			.ok()
-			.map(|recovered| Public(recovered.serialize_compressed()))
+			.map(|recovered| Public(recovered.serialize()))
 	}
 }
 
 #[cfg(feature = "full_crypto")]
-impl From<(secp256k1::Signature, secp256k1::RecoveryId)> for Signature {
-	fn from(x: (secp256k1::Signature, secp256k1::RecoveryId)) -> Signature {
+impl From<(secp256k1::Signature, secp256k1::recovery::RecoveryId)> for Signature {
+	fn from(x: (secp256k1::Signature, secp256k1::recovery::RecoveryId)) -> Signature {
 		let mut r = Self::default();
-		r.0[0..64].copy_from_slice(&x.0.serialize()[..]);
-		r.0[64] = x.1.serialize();
+		r.0[0..64].copy_from_slice(&x.0.serialize_compact()[..]);
+		r.0[64] = x.1.to_i32() as u8;
```

Comment:
This might cause issues in case of overflows - you can have multiple different `RecoveryId`s ending up with the same `Signature` type (malleability).

Seems that `libsecp256k1` was simply storing `u8`, but the C library stores `i32`.